src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                                    SyntaxFactory.Token(SyntaxKind.CloseBraceToken)),
+                                expressionBody: null);
+
+                            var localFunctionInvocation = SyntaxFactory.InvocationExpression(SyntaxFactory.IdentifierName(localFunctionName)).WithAdditionalAnnotations(Simplifier.Annotation);
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation.WithAdditionalAnnotations(Simplifier.Annotation));
+                            documentUpdateInfo = new DocumentUpdateInfo(parentStatement, new[] { localFunctionDeclaration, newParentExpressionStatement });
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (typeSymbol.TypeKind == TypeKind.Error ||
+                    typeSymbol.ContainsAnonymousType())
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private string GetFreeSymbolName(string prefix, int position)
+                => NameGenerator.GenerateUniqueName(
+                    prefix,
+                    _semanticModel.LookupSymbols(position).Select(symbol => symbol.Name).ToImmutableHashSet(),
+                    StringComparer.CurrentCulture);
+
+            private bool TryConvertIfInForEach(QueryExpressionSyntax source, ForEachStatementSyntax forEachStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before foreach(var x in from a in b select a)
+                var statement = forEachStatement.Statement;
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                {
+                    // before 
+                    //  foreach (var a in from a in b where a > 5 select a) 
+                    //  { 
+                    //      dosomething(a); 
+                    //  }
+                    // after 
+                    //  foreach (var a in b)
+                    //  {
+                    //      if (a > 5)
+                    //      {
+                    //          dosomething(a); 
+                    //      }
+                    //  }
+                    if (expression is IdentifierNameSyntax identifier &&
+                        identifier.Identifier.ValueText == forEachStatement.Identifier.ValueText)
+                    {
+                        return statement.WithAdditionalAnnotations(Formatter.Annotation);
```

Comment:
>return statement.WithAdditionalAnnotations(Formatter.Annotation); [](start = 24, length = 65)

Are we making an assumption that the types are the same? Why can we assume that?

Reply:
class A { }
    class B : A { }
...
foreach (A a in from B a in nums select a)

There would be a name conflict here. Suggesting to use TryReplaceWithLocalFunction for the case.



---
In reply to: [177169397](https://github.com/dotnet/roslyn/pull/25362#discussion_r177169397) [](ancestors = 177169397)